Working on improving inferability of sensitivity methods

chriselrod opened the pull request:
Not ready to merge yet; I'll make some more changes first.

Vaibhavdixit02 commented on src/global_sensitivity/sobol_sensitivity.jl:
```diff
@@ -22,57 +24,69 @@ function fuse_designs(A, B)
-             batch=false, Ei_estimator = :Jansen1999, kwargs...)
-    
+function gsa(f, method::Sobol, A::AbstractMatrix{TA}, B::AbstractMatrix{TB};
+             batch=false, Ei_estimator = :Jansen1999, kwargs...) where {TA,TB}
```
I don't think we'd need `TB` here

chriselrod replied:
Removed it.

chriselrod commented:
I checked a few code paths in the eFAST and Sobol methods, and all the type instabilities (from those paths) have been delayed until the final return value, at least given reasonable `f`. I split the functions into a before/after creation of `all_y` to provide a function barrier incase.

`QuasiMonteCarlo.generate_design_matrix` isn't type stable, but there's a function barrier.

Type instabilities of the final return are basically guaranteed in the current implementation because a few runtime variables determine return types, such as `batch::Bool` and `nboot`.

I removed the `Baes.@kwdef` because this blocked Revise.jl when changing anything in the same file (invalid redefinition of constant thanks to `gensym`).

Vaibhavdixit02 reviewed (approved):
Looks good to me.
